server/etcdserver/server.go

```diff
@@ -1901,6 +1901,11 @@ func (s *EtcdServer) applyEntryNormal(e *raftpb.Entry) {
 		s.w.Trigger(r.ID, s.applyV2Request((*RequestV2)(rp), shouldApplyV3))
 		return
 	}
+	if !shouldApplyV3 && raftReq.LeaseCheckpoint != nil {
```

Comment:
Problem with LeaseCheckpoints is that they are not persisted in backend at all. They only update the current in memory state. I think that forcing applying those entries is correct, but I'm not sure about current implementation.

Setting shouldApplyV3 in your code implies that checkpoint special case of V3, however I would argue that we should treat them separably. Instead of changing value of shouldApplyV3 I would change implementation of apply to move it above 
```
	if !shouldApplyV3 {
		return nil
	}
```

This way we ensure that ReleaseCheckpoint is always applied, however we make it explicit that checkpointing is not stored in backend.